### DISCUSSION TITLE
Annotating takes longer time

### DISCUSSION THREAD

**Discussion by tansifanzar** [explosion/spaCy#10172] [Help: Best practices]
I have a training dataset around .4M rows. When I am using the data for creating a training dataset for model, it is taking longer time inspite of using ProcessPoolExecutor. For example, when using nlp.pipe(text), it tkaes around 2minitues to process 500 rows. What is the optimum way to execute it in a shorter time?

**Comment by polm** [comment]
Please read the [speed FAQ](https://github.com/explosion/spaCy/discussions/8402). You can disable components you aren't using, or consider a smaller model.

**Reply by tansifanzar** [reply]
I have read the FAQ but disabling the other components except NER didn't solve the speed issue. Can you provide another suggestion?

**Reply by polm** [reply]
Hard to suggest much given just the information you have here. 

So you have 400,000 Docs. How long are these docs typically? When you say you're creating a training dataset using `nlp.pipe`, I don't understand what that means. Are you using the predictions of one model to create training data for another model? If so, why are you doing that? Are you using Transformers?

**Reply by tansifanzar** [reply]
1. Average length of docs is around 185
2. I am creating a training dataset using Entity Ruler providing patterns for different entities.

**Reply by polm** [reply]
OK. In your Entity Ruler, what kinds of patterns are you using? Are you using fixed string patterns, or are you matching on attributes? If matching on attributes, which ones? 

If you're just using phrase patterns (to match on literal text), then you don't need the NER component or a tok2vec when creating the Docs - you can just use a blank pipeline with an entity ruler (start with `spacy.blank("en")` or whatever instead of `spacy.load`). That would be a lot faster.

**Reply by tansifanzar** [reply]
I am using phrase patterns. I have done the way you mentioned but it's not faster. I have added a part of code here:

![image](https://user-images.githubusercontent.com/29386397/151792159-bf9ab411-5dda-47c3-b619-03f45a0060b5.png)

**Reply by pmbaumgartner** [reply]
> it is taking longer time inspite of using ProcessPoolExecutor

What do you mean by this? Is this what you're comparing against? How long do you think this _should_ take?

If you're trying to do multiprocessing with spaCy, rather than managing with `ProcessPoolExecutor`, you could instead try passing something to the `n_process` parameter in `nlp.pipe` - e.g. `nlp.pipe(texts, n_process=2)`. Start with 2 and if you see a speed up, you can increase that more.

---

Finally, as a general note, it makes it easier for everyone to help you if you include your code as a snippet of text [formatted as a code block](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) rather than an image, since people trying to help you will have to retype your entire code block from the image to try things out rather than just copy and paste it.

**Reply by tansifanzar** [reply]
I found out that ProcessPoolExecutor takes shorter time than using n_process in nlp_pipe. But according to my question, overall it takes longer which I want to solve. Sorry that previously I added a screenshot. Below I have added the code. 

```python
import concurrent.futures

nlp = spacy.blank("en")

ruler = nlp.add_pipe('entity_ruler')
ruler.add_patterns(final_pattern) # final_pattern is the dictionary of different entiities patterns


def testt(texts_fun):
    for doc in nlp.pipe([texts_fun]):
        #print(doc)
        spans = []
        for ent in doc.ents:
            span = {"start": ent.start_char, "end": ent.end_char, "label": ent.label_}
            spans.append(span)
        example = {"text": doc.text, "spans": spans}
        ex.append(example)
    return ex

with concurrent.futures.ProcessPoolExecutor(max_workers=6) as executor:
    result = executor.map(testt,texts) # texts corresponds to strings
    
 ```
